File: Entities/EntityManager.vb
Imports System.Runtime.CompilerServices

<Assembly: InternalsVisibleTo("Entities.Tests")>
''' <summary>
''' Tracks all entities in the game
''' </summary>
Public Class EntityManager
    Implements IEntityManager

    Private ReadOnly _entityMap As New Dictionary(Of Guid, BaseGameEntity)

    Public Sub RegisterEntity(ByRef newEntity As BaseGameEntity) Implements IEntityManager.RegisterEntity
        If newEntity IsNot Nothing Then
            _entityMap.Add(newEntity.Id, newEntity)
        End If
    End Sub

    Public Function GetEntityFromId(id As Guid) As BaseGameEntity Implements IEntityManager.GetEntityFromId
        If _entityMap.ContainsKey(id) Then
            Return _entityMap(id)
        End If
        Return Nothing
    End Function

    Public Sub RemoveEntity(ByRef pEntity As BaseGameEntity) Implements IEntityManager.RemoveEntity
        If pEntity IsNot Nothing AndAlso _entityMap.ContainsKey(pEntity.Id) Then
            _entityMap.Remove(pEntity.Id)
        End If
    End Sub

    Public Sub UpdateAll() Implements IEntityManager.UpdateAll
        If _entityMap.Count > 0 Then
            For Each e In _entityMap
                e.Value.Update()
            Next
        End If
    End Sub

    Public ReadOnly Property Count As Integer
        Get
            Return _entityMap.Count
        End Get
    End Property

    Public Function GetAllEntities() As List(Of BaseGameEntity) Implements IEntityManager.GetAllEntities
        If _entityMap.Count > 0 Then
            Return _entityMap.Values.ToList()
        End If
        Return New List(Of BaseGameEntity)
    End Function

End Class

File: space-game/Entities/EntityManager.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Runtime.CompilerServices;

[assembly: InternalsVisibleTo("Entities.Tests")]
namespace Entities
{
    /// <summary>
    /// Tracks all entities in the game
    /// </summary>
    public class EntityManager : IEntityManager
    {
        private readonly Dictionary<Guid, BaseGameEntity> _entityMap = new Dictionary<Guid, BaseGameEntity>();
        public void RegisterEntity(BaseGameEntity newEntity)
        {
            if (newEntity != null)
            {
                _entityMap.Add(newEntity.Id, newEntity);
            }
        }

        public BaseGameEntity GetEntityFromId(Guid id)
        {
            return _entityMap.ContainsKey(id) ? _entityMap[id] : null;
        }

        public void RemoveEntity(BaseGameEntity pEntity)
        {
            if (pEntity != null && _entityMap.ContainsKey(pEntity.Id))
            {
                _entityMap.Remove(pEntity.Id);
            }
        }

        public void UpdateAll()
        {
            if (_entityMap.Count > 0)
            {
                foreach (var e in _entityMap)
                {
                    e.Value.Update();
                }
            }
        }

        public int Count => _entityMap.Count;

        public List<BaseGameEntity> GetAllEntities()
        {
            return _entityMap.Count > 0 ? _entityMap.Values.ToList() : new List<BaseGameEntity>();
        }
    }
}
